src/test/java/com/puppycrawl/tools/checkstyle/filters/AbstractPatchFilterEvaluationTest.java

```diff
@@ -0,0 +1,106 @@
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.ConfigurationLoader;
+import com.puppycrawl.tools.checkstyle.ModuleFactory;
+import com.puppycrawl.tools.checkstyle.PackageObjectFactory;
+import com.puppycrawl.tools.checkstyle.PropertiesExpander;
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+import com.puppycrawl.tools.checkstyle.api.RootModule;
+import com.puppycrawl.tools.checkstyle.internal.utils.BriefUtLogger;
+
+abstract class AbstractPatchFilterEvaluationTest extends AbstractModuleTestSupport {
+
+    protected void testByConfig(String configPath)
+            throws Exception {
+        // we can add here any variable to provide path to patch name by PropertiesExpander
+        System.setProperty("tp", "src/test/resources/com/puppycrawl/tools/checkstyle/filters");
+        final Configuration config = ConfigurationLoader.loadConfiguration(
+                getPath(configPath),
+                new PropertiesExpander(System.getProperties()));
+        final ClassLoader moduleClassLoader = SuppressionPatchFilter.class.getClassLoader();
+        final ModuleFactory factory = new PackageObjectFactory(
+                SuppressionPatchFilter.class.getPackage().getName(), moduleClassLoader);
+
+        final RootModule rootModule = (RootModule) factory.createModule(config.getName());
+        rootModule.setModuleClassLoader(moduleClassLoader);
+        rootModule.configure(config);
+        final ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        rootModule.addListener(new BriefUtLogger(stream));
+
+        // run RootModule
+        final String inputFile = configPath.replaceFirst(
+                "(newline|patchedline)/(default|zero)ContextConfig.xml", "Test/");
+        final String path = getPath(inputFile);
+        final File file = new File(path);
+        final int errorCounter;
+
+        final List<File> theFiles;
+        if (file.isDirectory()) {
+            theFiles = Arrays.asList(file.listFiles());
+        }
+        else {
+            theFiles = Arrays.asList(file);
+        }
+        errorCounter = rootModule.process(theFiles);
+
+        // process each of the lines
+        try (ByteArrayInputStream inputStream =
+                     new ByteArrayInputStream(stream.toByteArray());
+             LineNumberReader lnr = new LineNumberReader(
+                     new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
+            final String expectedFile = configPath.replaceFirst(
+                    "(default|zero)ContextConfig.xml", "expected.txt");
+            final List<String> expected = Files.readAllLines(Paths.get(getPath(expectedFile)));
+            final List<String> actuals = lnr.lines().limit(expected.size()).sorted()
```

Comment:
at the beginning, I remove the sorted(), and CI failed, after I add it, it ok. Would the default sort be different on different machines?

Reply:
This sort is not good, it even made a lot of problems in main project.
We should make sort by numbers of  lines and column, to keep order natural.
Please try to fix it in separate PR if you find this inconvenient too

Reply:
We can use Google's Truth library to assert. It awesomely show diff in text.